[SourceKit] Add the raw doc comment to the cursor info response

ahoppen opened the pull request:
SourceKit-LSP currently parses the XML comment to generate Markdown again but round-tripping a (probably markdown) doc comment to XML to Markdown is lossy in many cases and unnecessary work. Include the comment as it is spelled in source in the cursor info response so that sourcekit-lsp can display it.

I did consider adding an option to the cursor info request that allows specification whether an XML or a raw comment should be returned but I didn’t want to slap yet another option to the cursor info request and the computation of the comment shouldn’t be very expensive. I’m open to discussion on this though.

Part of rdar://120685874

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on test/SourceKit/CursorInfo/cursor_info.swift:
```diff
@@ -307,6 +310,10 @@ let strInterpolation = "This is a \(stringStr + "ing") interpolation"
 // CHECK6-NEXT: <decl.function.free><syntaxtype.keyword>func</syntaxtype.keyword> <decl.name>fooSwiftFunc</decl.name>() -&gt; <decl.function.returntype><ref.struct usr="s:Si">Int</ref.struct></decl.function.returntype></decl.function.free>
+// CHECK6-NEXT: DOC COMMENT 
+// CHECK6-NEXT: This is 'fooSwiftFunc' from 'FooSwiftModule'. 
+// CHECK6-EMPTY: 
```
Does Clang return the newline too? Is this something we actually want?

bnbarham commented on lib/IDE/CommentConversion.cpp:
```diff
@@ -496,6 +496,33 @@ bool ide::getDocumentationCommentAsXML(const Decl *D, raw_ostream &OS,
+      return false;
+    }
+    clang::ASTContext &astCtx = CD->getASTContext();
+    const clang::RawComment *rawComment = astCtx.getRawCommentForAnyRedecl(CD);
```
Does the `FullComment` not have the raw comment? Seems like it would be nicer to share implementation between the two raw vs XML functions if we can. This also doesn't do `replaceObjcDeclarationsWithSwiftOnes`, which... maybe matters and maybe doesn't?

ahoppen replied:
Changed it to go through the same path that the old implementation did, which has some sophisticated logic in clang’s `ASTContext::getCommentForDecl`.

bnbarham reviewed (commented):
I'm fine with having it as well, though TBH I'm not sure anything is even using the XML comment any more. DocC will be using the symbol graph. I'm guessing we can probably just remove the XML entirely?

ahoppen committed:
[SourceKit] Add the raw doc comment to the cursor info response

SourceKit-LSP currently parses the XML comment to generate Markdown again but round-tripping a (probably markdown) doc comment to XML to Markdown is lossy in many cases and unnecessary work. Include the comment as it is spelled in source in the cursor info response so that sourcekit-lsp can display it.

Part of rdar://120685874

ahoppen commented:
@swift-ci Please smoke test

bnbarham commented on lib/IDE/CommentConversion.cpp:
```diff
@@ -496,6 +497,38 @@ bool ide::getDocumentationCommentAsXML(const Decl *D, raw_ostream &OS,
+    if (!FC) {
+      return false;
+    }
+    const clang::RawComment *rawComment = ClangContext.getRawCommentForAnyRedecl(FC->getDecl());
```
It's probably worth splitting that API up (similar to Swift's, eg. `getDocCommentProvidingDecl` -> the rest) so that we don't have to do this. But this is good enough for now 👍

bnbarham commented on lib/Markup/LineList.cpp:
```diff
@@ -34,8 +34,11 @@ std::string LineList::str() const {
   auto InitialIndentation = measureIndentation(FirstLine->Text);
 
   for (auto Line = FirstLine; Line != Lines.end(); ++Line) {
+    if (Line != FirstLine) {
```
Do you mean the end line here? Also, does the line have equality that just checks the range rather than the text itself? If not, might be better to just compare the range here rather than the text.

bnbarham replied:
Ah, true.
```
Stream << FirstLine->Text.drop_front(InitialIndentation);
for (auto *Line : Lines.drop_front()) {
  auto Drop = std::min(InitialIndentation, Line->FirstNonspaceOffset);
  Stream << "\n" << Line->Text.drop_front(Drop);
}
```
Though I didn't check what `Lines` is, so that may also need an `ArrayRef(Lines)`.

ahoppen replied:
Good suggestion. Added a variation of it. We still need to use C-style for loop because `FirstLine` might not be the first line, just the first line with content and we need to respect that.

bnbarham replied:
Oh that's a confusing name then 😅, thanks for updating. The above could just keep an index instead of bumping the pointer and then you could go for each, but this works too 👍

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test
